presto-orc/src/main/java/com/facebook/presto/orc/reader/ByteSelectiveStreamReader.java

```diff
@@ -456,4 +538,62 @@ public void close()
 
         systemMemoryContext.close();
     }
+
+    private boolean useBatchMode(int positionCount, int totalPositionCount)
+    {
+        // JMH benchmark shows that when there is no null or no filter, the batch read mode was always better than the skipping mode.
+        // When there is filter and partial nulls, the batch read mode was better than the skipping mode when the input filter rate is less than 0.4
+        if (presentStream == null || filter == null || (double) (totalPositionCount - positionCount) / totalPositionCount <= 0.4) {
+            return true;
+        }
+
+        return false;
+    }
+
+    private int evaluateFilter(int[] positions, int positionCount)
+    {
+        int positionsIndex = 0;
+        int i = 0;
+        while (i < positionCount) {
+            int position = positions[i];
+            if (filter.testLong(values[position])) {
+                outputPositions[positionsIndex++] = position;  // compact positions on the fly
+                i++;
+            }
+            else {
+                i += filter.getSucceedingPositionsToFail() + 1;
```

Comment:
why +1 ? is this code path coved by tests?

Reply:
the +1 is for the current position. For example, if an array has 8 elements (position 0 to 7) and position 1 fails, then getSucceedingPositionsToFail is 6, and getPrecedingPositionsToFail is 1. To skip to the next array (position 8), `i` need to be advanced with the following 6 elements and itself, therefore the +1. 

Array of integers are tested in existing test but there is no test for array of other types. I added  a new test for array of bytes in `testArrays()` in `TestSelectiveOrcReader`
```
       // non-null arrays of TINYINT
        tester.testRoundTrip(arrayType(TINYINT),
                createList(NUM_ROWS, i -> randomBytes(8, random)),
                ImmutableList.of(
                        ImmutableMap.of(new Subfield("c[2]"), IS_NULL),
                        ImmutableMap.of(new Subfield("c[2]"), nonNegative)));
```

Reply:
I get it now. I was confused because I missed that you are using while and not for loop. I think a for loop would be clearer.

```
        for (int i = 0; i < positionCount; i++) {
           int position = positions[i];
            if (filter.testLong(values[position])) {
                outputPositions[positionsIndex++] = position;  // compact positions on the fly
            }
            else {
                i += filter.getSucceedingPositionsToFail();
                positionsIndex -= filter.getPrecedingPositionsToFail();
            }
        }
```